sbi/analysis/plot.py

```diff
@@ -23,33 +23,427 @@
+            n_bins_row = int((limits_row[1] - limits_row[0]) / binsize_row)
+            kwargs['np_hist_kwargs']["bins"] = [n_bins_col, n_bins_row]
+        else:
+            # TODO: add more bin heuristics
+            pass
+    hist, xedges, yedges = np.histogram2d(
+        samples_col,
+        samples_row,
+        range=[
+            [limits_col[0], limits_col[1]],
+            [limits_row[0], limits_row[1]],
+        ],
+        **kwargs['np_hist_kwargs'],
+    )
+    ax.imshow(
+        hist.T,
+        extent=(
+            xedges[0],
+            xedges[-1],
+            yedges[0],
+            yedges[-1],
+        ),
+        **kwargs['mpl_kwargs'],
+    )
+
+
+def plt_kde_2d(
+    ax: Axes,
+    samples_col: np.ndarray,
+    samples_row: np.ndarray,
+    limits_col: torch.Tensor,
+    limits_row: torch.Tensor,
+    kwargs: Dict,
+) -> None:
+    """2D Kernel Density Estimation."""
+    density = gaussian_kde(
+        np.array([samples_col, samples_row]),
+        bw_method=kwargs["bw_method"],
+    )
+    X, Y = np.meshgrid(
+        np.linspace(
+            limits_col[0],
+            limits_col[1],
+            kwargs["bins"],
+        ),
+        np.linspace(
+            limits_row[0],
+            limits_row[1],
+            kwargs["bins"],
+        ),
+    )
+    positions = np.vstack([X.ravel(), Y.ravel()])
+    Z = np.reshape(density(positions).T, X.shape)
+
+    ax.imshow(
+        Z,
+        extent=(
+            limits_col[0],
+            limits_col[1],
+            limits_row[0],
+            limits_row[1],
+        ),
+        **kwargs['mpl_kwargs'],
+    )
+
+
+def plt_contour_2d(
+    ax: Axes,
+    samples_col: np.ndarray,
+    samples_row: np.ndarray,
+    limits_col: torch.Tensor,
+    limits_row: torch.Tensor,
+    kwargs: Dict,
+) -> None:
+    """2D Contour based on Kernel Density Estimation."""
+    density = gaussian_kde(
+        np.array([samples_col, samples_row]),
+        bw_method=kwargs["bw_method"],
+    )
+    X, Y = np.meshgrid(
+        np.linspace(
+            limits_col[0],
+            limits_col[1],
+            kwargs["bins"],
+        ),
+        np.linspace(
+            limits_row[0],
+            limits_row[1],
+            kwargs["bins"],
+        ),
+    )
+    positions = np.vstack([X.ravel(), Y.ravel()])
+    Z = np.reshape(density(positions).T, X.shape)
+    if kwargs["percentile"]:
+        Z = probs2contours(Z, kwargs["levels"])
+    else:
+        Z = (Z - Z.min()) / (Z.max() - Z.min())
+
+    ax.contour(
+        X,
+        Y,
+        Z,
+        extent=(
+            limits_col[0],
+            limits_col[1],
+            limits_row[0],
+            limits_row[1],
+        ),
+        levels=kwargs["levels"],
+        **kwargs['mpl_kwargs'],
+    )
+
+
+def plt_scatter_2d(
+    ax: Axes,
+    samples_col: np.ndarray,
+    samples_row: np.ndarray,
+    limits_col: torch.Tensor,
+    limits_row: torch.Tensor,
```

Comment:
args not used. if the signature requires having them, maybe add a separate `**kwargs` to absorb them and call the kwarg `kwargs` differently?

Reply:
My reasoning was that almost all plotting functions use these arguments, so it might be more readable to be consistent like this? But if you prefer we could absorb  them into **kwargs.

Reply:
OK, I think it's better to not have them / absorb them into the `kwargs`.

Reply:
we decided to keep them to have a common API for plotting methods.